esp.c

```diff
@@ -31,40 +31,93 @@
+
+#define SEMAPHORE_TIMEOUT_MS 50
+
+  static rmt_data_t *led_data = NULL;
+  static uint32_t led_data_size = 0;
+  static int rmtPin = -1;
+
+  if (show_mutex && xSemaphoreTake(show_mutex, SEMAPHORE_TIMEOUT_MS / portTICK_PERIOD_MS) == pdTRUE) {
+    uint32_t requiredSize = numBytes * 8;
+    if (requiredSize > led_data_size) {
+      free(led_data);
+      if (led_data = (rmt_data_t *)malloc(requiredSize * sizeof(rmt_data_t))) {
+        led_data_size = requiredSize;
+      } else {
+        led_data_size = 0;
+      }
+    } else if (requiredSize == 0) {
+      // To release RMT resources (RMT channels and led_data), call
+      //  .updateLength(0) to set number of pixels/bytes to zero,
+      //  then call .show() to invoke this code and free resources.
+      free(led_data);
+      led_data = NULL;
+      if (rmtPin >= 0) {
+        rmtDeinit(rmtPin);
+        rmtPin = -1;
+      }
+      led_data_size = 0;
+    }
 
-  if (!rmtInit(pin, RMT_TX_MODE, RMT_MEM_NUM_BLOCKS_1, 10000000)) {
-    log_e("Failed to init RMT TX mode on pin %d", pin);
-    return;
-  }
+    if (led_data_size > 0 && requiredSize <= led_data_size) {
+      if (pin != rmtPin) {
+        if (rmtPin >= 0) {
+          rmtDeinit(rmtPin);
+          rmtPin = -1;
+        }
+        if (!rmtInit(pin, RMT_TX_MODE, RMT_MEM_NUM_BLOCKS_1, 10000000)) {
+          log_e("Failed to init RMT TX mode on pin %d", pin);
+          return;
+        }
+        rmtPin = pin;
+      }
 
-  int i=0;
-  for (int b=0; b < numBytes; b++) {
-    for (int bit=0; bit<8; bit++){
-      if ( pixels[b] & (1<<(7-bit)) ) {
-        led_data[i].level0 = 1;
-        led_data[i].duration0 = 8;
-        led_data[i].level1 = 0;
-        led_data[i].duration1 = 4;
-      } else {
-        led_data[i].level0 = 1;
-        led_data[i].duration0 = 4;
-        led_data[i].level1 = 0;
-        led_data[i].duration1 = 8;
+      if (rmtPin >= 0) {
+        int i=0;
+        for (int b=0; b < numBytes; b++) {
+          for (int bit=0; bit<8; bit++){
+            if ( pixels[b] & (1<<(7-bit)) ) {
+              led_data[i].level0 = 1;
+              led_data[i].duration0 = 8;
+              led_data[i].level1 = 0;
+              led_data[i].duration1 = 4;
+            } else {
+              led_data[i].level0 = 1;
+              led_data[i].duration0 = 4;
+              led_data[i].level1 = 0;
+              led_data[i].duration1 = 8;
+            }
+            i++;
+          }
+        }
+
+        rmtWrite(pin, led_data, numBytes * 8, RMT_WAIT_FOR_EVER);
       }
-      i++;
     }
-  }
 
-  //pinMode(pin, OUTPUT);  // don't do this, will cause the rmt to disable!
-  rmtWrite(pin, led_data, numBytes * 8, RMT_WAIT_FOR_EVER);
+    xSemaphoreGive(show_mutex);
+  }
 }
 
-
+// To avoid race condition initializing the mutex, all instances of
+//  Adafruit_NeoPixel must be constructed before launching and child threads
+void espInit() {
+  if (!show_mutex) {
+    show_mutex = xSemaphoreCreateMutex();
```

Comment:
This would be a perfect solution if it compiled; unfortunately this isn't a c++ module so initializing statics with non-constants is not supported.

Reply:
Uff, I forgot this was a  `.c` file.

The `xSemaphoreCreateMutexStatic` is only different in that it uses a user-provided buffer instead of allocating a buffer, I don't think it's really "static" in the sense of being able to do static initalization unfortunately.